src/Parser/DestructionWarlock/Modules/SoulShards/SoulShardEvents.js

```diff
@@ -0,0 +1,254 @@
+  [SPELLS.GRIMOIRE_FELHUNTER.id]: SUMMON_COST,
+  [SPELLS.SUMMON_VOIDWALKER.id]: SUMMON_COST,
+  [SPELLS.SUMMON_SUCCUBUS.id]: SUMMON_COST,
+  [SPELLS.SUMMON_FELHUNTER.id]: SUMMON_COST,
+};
+
+const FRAGMENT_GENERATING_ENERGIZE_ABILITIES = [
+  SPELLS.SOUL_CONDUIT_SHARD_GEN.id,
+  SPELLS.SOULSNATCHER_FRAGMENT_GEN.id,
+  SPELLS.FERETORY_OF_SOULS_FRAGMENT_GEN.id,
+];
+
+class SoulShardEvents extends Module {
+  static dependencies = {
+    enemies: Enemies,
+    combatants: Combatants,
+  };
+
+  hasT20_2p = false;
+  currentFragments = 0;
+
+  on_initialized() {
+    if (!this.owner.error) {
+      this.hasT20_2p = this.combatants.selected.hasBuff(SPELLS.WARLOCK_DESTRO_T20_2P_BONUS.id);
+      this.currentFragments = 30; //on the start of the fight we should have 3 soul shards (30 fragments) by default
+      debug && console.log("start fragments " + this.currentFragments);
+    }
+  }
+
+  on_byPlayer_energize(event) {
+    if (event.resourceChangeType !== RESOURCE_TYPES.SOUL_SHARDS) {
+      return;
+    }
+    const spellId = event.ability.guid;
+    if (FRAGMENT_GENERATING_ENERGIZE_ABILITIES.indexOf(spellId) === -1) {
+      return;
+    }
+
+    const shardEvent = {
+      timestamp: event.timestamp,
+      type: 'soulshardfragment_gained',
+      ability: {
+        guid: spellId,
+        name: SPELLS[spellId].name,
+      },
+    };
+    //purposefully missing targetID and targetInstance because both source and target on energize is the player
+
+    const rawGain = event.resourceChange * FRAGMENTS_PER_SHARD;
+    let gain = 0;
+    let waste = 0;
+    if (this.currentFragments + rawGain > MAX_FRAGMENTS) {
+      gain = MAX_FRAGMENTS - this.currentFragments;
+      waste = this.currentFragments + rawGain - MAX_FRAGMENTS;
+    }
+    else {
+      gain = rawGain;
+    }
+
+    this.currentFragments += gain;
+
+    shardEvent.amount = gain;
+    shardEvent.waste = waste;
+    shardEvent.currentFragments = this.currentFragments;
+
+    debug && console.log('++ ' + shardEvent.amount + '(w: ' + shardEvent.waste + ') = ' + shardEvent.currentFragments + ', ' + shardEvent.ability.name + ', orig: ', event);
+    this.owner.triggerEvent('soulshardfragment_gained', shardEvent);
+  }
+
+  //should also work with Havoc and talent Fire and Brimstone
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if (FRAGMENT_GENERATING_ABILITIES.indexOf(spellId) === -1) {
+      return;
+    }
+
+    const shardEvent = {
+      timestamp: event.timestamp,
+      type: 'soulshardfragment_gained',
+      ability: {
+        guid: spellId,
+        name: SPELLS[spellId].name,
+      },
+      targetID: event.targetID,
+      targetInstance: event.targetInstance,
+    };
+    let rawGain;
+    let gain = 0;
+    let waste = 0;
+
+    switch (spellId) {
+      case SPELLS.IMMOLATE_DEBUFF.id:
+        rawGain = 1;
+        break;
+      case SPELLS.CONFLAGRATE.id:
+        rawGain = 5;
+        break;
+      case SPELLS.INCINERATE.id:
```

Comment:
This changes things up a little from the previous suggestion:

```javascript
const SHARD_GENERATION = {
  [SPELLS.CONFLAGRATE.id]: () => 5,
  [SPELLS.INCINERATE.id]: () => {
    // Return shard generation here
  },
  ...
};

const rawGain = SHARD_GENERATION[spellid]();
```

Reply:
Yes, exactly this I had in mind, I just don't know if I could access the 2 dependent modules from inside the function (I need to look for T20 2p buff on the player, Havoc debuff on target and also if the event (if it critted) in order to determine how many fragments to add).

Reply:
These short hand functions `() => x` will be bound to the parent scope here instead of `SHARD_GENERATION`, if the syntax was:

```javascript
const SHARD_GENERATION = {
  [SPELLS.INCINERATE.id]: function() {
    this.module...
  }
};
```
your concern would be entirely correct, as `this` will refer to the object you've defined it in.

---
I'd recommend [YDKJS - Scope and Closures](https://github.com/getify/You-Dont-Know-JS/tree/master/scope%20%26%20closures) as some reading if you want to become certain on these types of thoughts.

Reply:
Ah, I see.. but I still don't have an information about critting.. but I suppose I could do something like this:
`
const SHARD_GENERATION = {
   [SPELLS.CONFLAGRATE.id]: (isCrit) => 5,
   [SPELLS.INCINERATE.id]: (isCrit) => {
   },
   ...
};

const rawGain = SHARD_GENERATION[spellId](event.hitType);
`

most of the time it wouldn't be used but that doesn't matter very much, no?

Reply:
`event` will be accessible inside this scope, so you shouldn't have a problem with that.